File: ios/SmashCore/Sources/SmashCore/Feel/SoundMix.swift
import Foundation

/// How a play of an event is mixed (spec §8.8, shared/data/sounds.toml) — the same arithmetic on both
/// platforms' playback layers. A sound made *somewhere on the pitch* (it carries an x) follows §8.6's
/// time scale and pans with where it happened; every other sound plays at real speed, centred.
public enum SoundMix {
    /// dB → linear amplitude.
    public static func amplitude(_ db: Double) -> Double { pow(10, db / 20) }

    /// A play's pitch factor for a uniform draw `u` in [0, 1): ± `semitones`, uniformly in semitones.
    public static func pitch(semitones: Double, u: Double) -> Double { pow(2, (2 * u - 1) * semitones / 12) }

    /// A play's rate: its pitch, times the time scale when it happened on the pitch, clamped to the
    /// playback layer's range.
    public static func rate(pitch: Double, timeScale: Double, onPitch: Bool, min lo: Double, max hi: Double) -> Double {
        Swift.min(Swift.max(onPitch ? pitch * timeScale : pitch, lo), hi)
    }

    /// Where a sound at `x` across the pitch (half-width 15) sits: −1 left … 1 right, `width` at the
    /// boards; centred when it has no place.
    public static func pan(x: Double?, width: Double) -> Double {
        guard let x else { return 0 }
        return Swift.min(Swift.max(x / Tuning.Pitch.halfWidth * width, -1), 1)
    }

    /// Left and right gains for `pan` under a constant-power law (pan 0: both 1/√2).
    public static func stereo(_ pan: Double) -> (left: Double, right: Double) {
        let a = (pan + 1) * Double.pi / 4
        return (cos(a), sin(a))
    }

    /// Which of `count` variants a play takes for a uniform draw `u` in [0, 1): at random, never the
    /// one played last (`last`) when there are two or more.
    public static func variant(count: Int, last: Int?, u: Double) -> Int {
        guard count > 1 else { return 0 }
        guard let last, last >= 0, last < count else { return Swift.min(Int(u * Double(count)), count - 1) }
        let pick = Swift.min(Int(u * Double(count - 1)), count - 2)
        return pick >= last ? pick + 1 : pick
    }
}

/// The voices a playback layer has (spec §8.8): which one a new play takes. An event already at its
/// `maxVoices` steals its own oldest; otherwise a free voice; otherwise the oldest voice of the lowest
/// priority below the event's; otherwise the play is dropped.
public struct VoicePool: Sendable {
    public struct Voice: Sendable, Hashable {
        public let cue: SoundCue
        public let priority: Int
        public let started: Double
        public let ends: Double
    }

    public private(set) var voices: [Voice?]

    public init(count: Int) { voices = Array(repeating: nil, count: count) }

    /// The voice a play of `cue` starting at `now` and lasting `seconds` takes, or nil to drop it.
    /// The voice is recorded as taken.
    public mutating func claim(_ cue: SoundCue, now: Double, seconds: Double) -> Int? {
        let spec = cue.spec
        func live(_ i: Int) -> Voice? { voices[i].flatMap { $0.ends > now ? $0 : nil } }
        let chosen: Int?
        let mine = voices.indices.filter { live($0)?.cue == cue }
        if mine.count >= spec.maxVoices {
            chosen = mine.min { voices[$0]!.started < voices[$1]!.started }
        } else if let free = voices.indices.first(where: { live($0) == nil }) {
            chosen = free
        } else {
            chosen = voices.indices
                .filter { voices[$0]!.priority < spec.priority }
                .min { a, b in
                    let va = voices[a]!, vb = voices[b]!
                    return va.priority != vb.priority ? va.priority < vb.priority : va.started < vb.started
                }
        }
        guard let i = chosen else { return nil }
        voices[i] = Voice(cue: cue, priority: spec.priority, started: now, ends: now + seconds)
        return i
    }
}

File: android/core/src/main/kotlin/in/nann/smashhockey/core/feel/SoundMix.kt
package `in`.nann.smashhockey.core.feel

import `in`.nann.smashhockey.core.generated.SoundCue
import `in`.nann.smashhockey.core.generated.Tuning
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin

/**
 * How a play of an event is mixed (spec §8.8, shared/data/sounds.toml) — the same arithmetic as iOS's
 * playback layer. A sound made *somewhere on the pitch* (it carries an x) follows §8.6's time scale
 * and pans with where it happened; every other sound plays at real speed, centred.
 */
object SoundMix {
    /** dB → linear amplitude. */
    fun amplitude(db: Double): Double = 10.0.pow(db / 20)

    /** A play's pitch factor for a uniform draw [u] in [0, 1): ± [semitones], uniformly in semitones. */
    fun pitch(semitones: Double, u: Double): Double = 2.0.pow((2 * u - 1) * semitones / 12)

    /** A play's rate: its pitch, × the time scale when it happened on the pitch, clamped to the layer's range. */
    fun rate(pitch: Double, timeScale: Double, onPitch: Boolean, min: Double, max: Double): Double =
        (if (onPitch) pitch * timeScale else pitch).coerceIn(min, max)

    /** Where a sound at [x] across the pitch (half-width 15) sits: −1 left … 1 right, [width] at the boards; centred without a place. */
    fun pan(x: Double?, width: Double): Double {
        if (x == null) return 0.0
        return (x / Tuning.Pitch.halfWidth * width).coerceIn(-1.0, 1.0)
    }

    /** Left and right gains for [pan] under a constant-power law (pan 0: both 1/√2). */
    fun stereo(pan: Double): Pair<Double, Double> {
        val a = (pan + 1) * PI / 4
        return cos(a) to sin(a)
    }

    /**
     * Which of [count] variants a play takes for a uniform draw [u] in [0, 1): at random, never the one
     * played last ([last]) when there are two or more.
     */
    fun variant(count: Int, last: Int?, u: Double): Int {
        if (count <= 1) return 0
        if (last == null || last < 0 || last >= count) return minOf((u * count).toInt(), count - 1)
        val pick = minOf((u * (count - 1)).toInt(), count - 2)
        return if (pick >= last) pick + 1 else pick
    }
}

/**
 * The voices a playback layer has (spec §8.8): which one a new play takes. An event already at its
 * `maxVoices` steals its own oldest; otherwise a free voice; otherwise the oldest voice of the lowest
 * priority below the event's; otherwise the play is dropped.
 */
class VoicePool(count: Int) {
    data class Voice(val cue: SoundCue, val priority: Int, val started: Double, val ends: Double)

    private val slots = arrayOfNulls<Voice>(count)
    val voices: List<Voice?> get() = slots.toList()

    /** The voice a play of [cue] starting at [now] and lasting [seconds] takes, or null to drop it. */
    fun claim(cue: SoundCue, now: Double, seconds: Double): Int? {
        val spec = cue.spec
        fun live(i: Int) = slots[i]?.takeIf { it.ends > now }
        val mine = slots.indices.filter { live(it)?.cue == cue }
        val chosen: Int? = if (mine.size >= spec.maxVoices) {
            mine.minByOrNull { slots[it]!!.started }
        } else {
            slots.indices.firstOrNull { live(it) == null }
                ?: slots.indices.filter { slots[it]!!.priority < spec.priority }
                    .minWithOrNull(compareBy<Int>({ slots[it]!!.priority }, { slots[it]!!.started }))
        }
        val i = chosen ?: return null
        slots[i] = Voice(cue, spec.priority, now, now + seconds)
        return i
    }
}
